client/src/api/ApiClient.ts

```diff
@@ -1,28 +1,73 @@
+import { setCookie } from '../SetCookie';
+import { ApiError, ApiErrorFromResponse } from './ApiError';
+
 export interface ApiClient {
-  authenticate(username: string, password: string): Promise<boolean>
+  authenticate(username: string, password: string): Promise<boolean | ApiError>
+  authenticated(): Promise<boolean | ApiError>
 }
 
-// TODO: Add CSRF protection
-export class ApiClientImpl {
+export class ApiClientImpl implements ApiClient {
   private baseRoute = '/api';
 
-  authenticate(username: string, password: string): Promise<boolean> {
-    return fetch(`${this.baseRoute}/authenticate`, {
-      method: 'POST',
-      cache: 'no-cache',
-      headers: {
-        'Content-Type': 'application/json',
-      },
-      body: JSON.stringify({
+  private readonly csrf: string;
+
+  constructor() {
+    const randomNumbers = new Uint32Array(8);
+    crypto.getRandomValues(randomNumbers);
+    this.csrf = encodeURIComponent(btoa(randomNumbers.join('')));
+    setCookie('csrf-token', this.csrf, 6);
```

Comment:
My understanding is that for double submit the important aspect is that a cookie value and header value set to some cryptographically strong value match on the server side. In the `makeCall` method I also include the csrf token as a header and verify that they both match server side. Here's what I'm referencing: 

https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie

Admittedly I've never hand-rolled CSRF protection and would love to learn what's missing.

Reply:
Typically the server will generate and send the cookie to the client, then the client submits that cookie. I'll look at your submission in more detail tomorrow morning.

Reply:
The cookie needs to be set by the server and client needs to send it back with each http request. This way when a user clicks on URL that points to some API action the CSRF protection will kick in. Current solution wont work with sub-domains because it's possible to set a cookie from javascript for parent domains. For example, a user can visit a subdomain website which will send API requests to your endpoints on user behalf.